### PR TITLE
Docs: Add plugin migration guide

### LINKED ISSUE #22427: Migration guide for plugins
Hi Team ,

I am reading the document for the developing the graphan plugin but I could see the documentation is not proper meaning that no detail information is provided . can you please add some more documentation on how the folder structure, which are files important , what each file does ..etc then how to build the plugin , what are the classes required .. and if possible with good examples .. I still didn't get the starting point after reading this documentation.

Thanks
Sanket

### PR DESCRIPTION
**Pull request by marcusolsson**
**What this PR does / why we need it**:

Adds a migration guide to the plugin docs to help plugin authors when migrating their plugin to the new React platform.

Fixes #22427

### CODE REVIEW

File: docs/sources/developers/plugins/migration-guide.md [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+```
+
+Would now be returning a data frame with one field of type time, and one with type number:
+
+```
+async query(options: DataQueryRequest<MyQuery>): Promise<DataQueryResponse> {
+  return {
+    data: options.targets.map(query => {
+      return new MutableDataFrame({
+        refId: query.refId,
```
Comment (yesoreyeram) [inline docs/sources/developers/plugins/migration-guide.md:130]: It is bit unclear where the seriesname `upper_75` goes in the new format.

**Review by kminehart — DISMISSED** [review]
Just one minor change :D

Looks great!

File: docs/sources/developers/plugins/migration-guide.md [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+
+This guide is aimed towards plugin authors who want to migrate their plugins to Grafana 7.0, and users who are using custom plugins and want to know whether it’s safe to upgrade.
+
+## What's new in Grafana 7?
+
+Grafana 7 introduced a whole new plugin platform based on React that supersedes the previous Angular based platform. Plugins built using Angular will still work for the foreseeable future, though we encourage new plugin authors to develop against the new platform.
+
+### New data format
+
+Along with the move to React, the new plugin platform introduced a new internal data format, called data frames. Previously, data source plugins could send data either as time series or tables. With data frames, data sources can send any type of data in a table-like structure. This gives you more flexibility for data you want to visualize in Grafana.
```
Comment (dprokop) [inline docs/sources/developers/plugins/migration-guide.md:16]:
```suggestion
Along with the move to React, the new plugin platform introduced a new internal data format, called [data frames](https://grafana.com/docs/grafana/latest/developers/plugins/data-frames/#data-frames). Previously, data source plugins could send data either as time series or tables. With data frames, data sources can send any type of data in a table-like structure. This gives you more flexibility for data you want to visualize in Grafana.
```

File: docs/sources/developers/plugins/migration-guide.md [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+
+While the previous Angular-based plugin SDK did support TypeScript, for the React platform, we’ve bet hard on TypeScript. All our APIs are now TypeScript, which might require existing code to update to the new stricter requirements. Grafana 7 also introduced several new APIs for plugin developers that take advantage of many of the new features in Grafana 7.
+
+### Grafana Toolkit
+
+With Grafana 7.0, we released a new tool for making it easier to develop plugins. Before, you’d use Gulp, Grunt, or similar tools to generate the minified assets.
+
+### Field options
+
+Grafana 7 introduced the concept of _field options_, a new way of configuring your data before it gets visualized. Since this was not available in previous versions, any plugin that enabled field-based configuration will not work in previous versions of Grafana.
```
Comment (dprokop) [inline docs/sources/developers/plugins/migration-guide.md:28]:
```suggestion
Grafana 7 introduced the concept of [_field options_](https://grafana.com/docs/grafana/latest/panels/field-configuration-options/#field-configuration-options), a new way of configuring your data before it gets visualized. Since this was not available in previous versions, any plugin that enabled field-based configuration will not work in previous versions of Grafana.
```

**Review by dprokop — APPROVED** [review]
Great work @marcusolsson!

File: docs/sources/developers/plugins/migration-guide.md [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+
+While the previous Angular-based plugin SDK did support TypeScript, for the React platform, we’ve bet hard on TypeScript. All our APIs are now TypeScript, which might require existing code to update to the new stricter requirements. Grafana 7 also introduced several new APIs for plugin developers that take advantage of many of the new features in Grafana 7.
+
+### Grafana Toolkit
+
+With Grafana 7.0, we released a new tool for making it easier to develop plugins. Before, you’d use Gulp, Grunt, or similar tools to generate the minified assets.
+
+### Field options
+
+Grafana 7 introduced the concept of [_field options_](https://grafana.com/docs/grafana/latest/panels/field-configuration-options/#field-configuration-options), a new way of configuring your data before it gets visualized. Since this was not available in previous versions, any plugin that enabled field-based configuration will not work in previous versions of Grafana.
```
Comment (oddlittlebird) [inline docs/sources/developers/plugins/migration-guide.md:28]: This link needs to be a relref

Comment (oddlittlebird) [inline docs/sources/developers/plugins/migration-guide.md:28]:
```suggestion
Grafana 7 introduced the concept of [_field options_](https://grafana.com/docs/grafana/latest/panels/field-configuration-options/#field-configuration-options), a new way of configuring your data before it gets visualized. Since this was not available in previous versions, any plugin that enables field-based configuration will not work in previous versions of Grafana.
```

File: docs/sources/developers/plugins/migration-guide.md [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+
+For plugins prior to Grafana 7.0, all options are considered Display options. The tab for field configuration will then be hidden.
```
Comment (oddlittlebird) [inline docs/sources/developers/plugins/migration-guide.md:30]:
```suggestion
For plugins prior to Grafana 7.0, all options are considered _Display options_. The tab for field configuration is not available.
```

File: docs/sources/developers/plugins/migration-guide.md [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+
+### Backend plugins
+
+While backend plugins were available as an experimental feature in previous versions of Grafana, the support has been greatly improved for Grafana 7. Backend plugins for Grafana 7 are backwards compatible and will continue to work. However, the old backend plugin system has been deprecated, and we recommend that you use the new SDK for backend plugins.
```
Comment (oddlittlebird) [inline docs/sources/developers/plugins/migration-guide.md:34]:
```suggestion
While backend plugins were available as an experimental feature in previous versions of Grafana, the support has been greatly improved for Grafana 7. Backend plugins for Grafana 7 are backwards-compatible and will continue to work. However, the old backend plugin system has been deprecated, and we recommend that you use the new SDK for backend plugins.
```

File: docs/sources/developers/plugins/migration-guide.md [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+
+Since Grafana 7 introduced [signing of backend plugins](https://grafana.com/docs/grafana/latest/plugins/plugin-signature-verification/), community plugins won’t load by default if they’re unsigned.
```
Comment (oddlittlebird) [inline docs/sources/developers/plugins/migration-guide.md:36]: This link sould be a relref. Apply throughout.

File: docs/sources/developers/plugins/migration-guide.md [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+
+To learn more, refer to [Backend plugins](https://grafana.com/docs/grafana/latest/developers/plugins/backend/).
+
+## Why should I migrate my plugin?
+
+There are several benefits in using the new plugin platform.
+
+- **Better performance:** Components written in React are more responsive.
+- **Field options**: By migrating to the new data frame format, you can leverage the new field options to let users fine-grained customization of the data and display options.
```
Comment (oddlittlebird) [inline docs/sources/developers/plugins/migration-guide.md:45]:
```suggestion
- **Field options:** By migrating to the new data frame format, you can leverage the new field options to let users customize their data and display options.
```

File: docs/sources/developers/plugins/migration-guide.md [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+
+For example:
+
+- A legacy panel with data source that returns data frames: Grafana converts the response to the legacy format.
+- A legacy data source with a panel using data frames: Grafana converts the response to the data frame format.
+- If both panel and data source uses the same format, no transformations are made. Data is passed as is.
+
+### target and jsonData are unchanged
+
+The query model, `target`, and the configuration model, jsonData,  are still the same. This means that if you use the same query model and configuration for your plugin, the migrated plugin will use existing queries and configuration. You don’t have to worry about breaking existing dashboards.
```
Comment (oddlittlebird) [inline docs/sources/developers/plugins/migration-guide.md:63]:
```suggestion
The query model, `target`, and the configuration model, jsonData, are still the same. This means that if you use the same query model and configuration for your plugin, then the migrated plugin will use existing queries and configuration. You don’t have to worry about breaking existing dashboards.
```

File: docs/sources/developers/plugins/migration-guide.md [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+
+## How do I migrate from an Angular plugin to a React plugin?
+
+If you’re looking to migrate a plugin to the new plugin platform, we recommend that you release it under a new major version, and that you keep a release branch for the previous version to be able to roll out patch releases for versions prior to Grafana 7.
```
Comment (oddlittlebird) [inline docs/sources/developers/plugins/migration-guide.md:67]:
```suggestion
If you’re looking to migrate a plugin to the new plugin platform, then we recommend that you release it under a new major version, and that you keep a release branch for the previous version to be able to roll out patch releases for versions prior to Grafana 7.
```

Comment (oddlittlebird) [reply docs/sources/developers/plugins/migration-guide.md:67]: Best practice is to keep sentences to 25 words or less. Please break this up.

File: docs/sources/developers/plugins/migration-guide.md [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+
+While there’s no 1-to-1 migration path from a Angular plugin to the new React platform, from early adopters, we’ve learned that one of the easiest ways to migrate is to create a new plugin from one of the templates provided by Grafana Toolkit, and then move the existing code into the new plugin.
```
Comment (oddlittlebird) [inline docs/sources/developers/plugins/migration-guide.md:69]:
```suggestion
While there’s no 1-to-1 migration path from an Angular plugin to the new React platform, from early adopters, we’ve learned that one of the easiest ways to migrate is to create a new plugin from one of the templates provided by Grafana Toolkit, and then move the existing code into the new plugin.
```

Comment (oddlittlebird) [reply docs/sources/developers/plugins/migration-guide.md:69]: Break up this sentence as well.

File: docs/sources/developers/plugins/migration-guide.md [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+
+An example of how you could do this is to create a new branch called `migrate-to-react`, delete everything and start with one of the plugin templates.
```
Comment (oddlittlebird) [inline docs/sources/developers/plugins/migration-guide.md:71]:
```suggestion
An example of how you could do this is to create a new branch called `migrate-to-react`, delete everything, and then start with one of the plugin templates.
```

File: docs/sources/developers/plugins/migration-guide.md [hunk 13]
```diff
@@ -0,0 +1,151 @@
...
+
+### Migrate a panel plugin
+
+In <7.0, you would export a MetricsPanelCtrl from module.ts.
```
Comment (oddlittlebird) [inline docs/sources/developers/plugins/migration-guide.md:75]:
```suggestion
Prior to Grafana 7.0, you would export a MetricsPanelCtrl from module.ts.
```

File: docs/sources/developers/plugins/migration-guide.md [hunk 14]
```diff
@@ -0,0 +1,151 @@
...
+import { MetricsPanelCtrl } from 'grafana/app/plugins/sdk';
+
+class MyPanelCtrl extends MetricsPanelCtrl {
+  // ...
+}
+
+export { MyPanelCtrl as PanelCtrl };
+```
+
+In >=7.0, plugins now export a PanelPlugin from module.ts
```
Comment (oddlittlebird) [inline docs/sources/developers/plugins/migration-guide.md:87]: The less than or equal to symbols in prose are very hard to parse. Please do not use them. Use words like "before" instead.

File: docs/sources/developers/plugins/migration-guide.md [hunk 15]
```diff
@@ -0,0 +1,151 @@
...
+
+```ts
+import { PanelPlugin } from '@grafana/data';
+
+export const plugin = new PanelPlugin<MyOptions>(MyPanel);
+```
+
+where MyPanel is a React component containing the props from PanelProps.
```
Comment (oddlittlebird) [inline docs/sources/developers/plugins/migration-guide.md:95]: Please make this a complete sentence, or at least make it more understandable on its own. Also, sentence case.

File: docs/sources/developers/plugins/migration-guide.md [hunk 16]
```diff
@@ -0,0 +1,151 @@
...
+export const MyPanel: React.FC<Props> = ({ options, data, width, height }) => {
+  // ...
+}
+```
+
+### Migrate a data source plugin
+
+For many data sources, the `ConfigEditor` is the simplest component, so it’s a good candidate to start migrating.
+
+1. When you’ve defined your configuration model, the next step is to implement the `testDatasource()` method on the class that extends `DataSourceApi`.
```
Comment (oddlittlebird) [inline docs/sources/developers/plugins/migration-guide.md:111]: Are these steps? I don't understand what you are telling me to do in any of them. Please use first person imperative case rather than this indirect verbiage.

File: docs/sources/developers/plugins/migration-guide.md [hunk 17]
```diff
@@ -0,0 +1,151 @@
...
+1. The `query()` method supports both the new data frame response and the old TimeSeries response, so don’t worry about converting to the new format just yet.
+1. Next, implement the `QueryEditor`. How much work this requires depends on how complex your query model is.
+
+By now, you should be able to release your new version.
+
+To fully migrate to the new plugin platform, convert the time series response to a data frame response.
+
+### Migrate to data frames
+
+In <7.0, data source and panel plugins exchanged data using either time series or tables. In >=7.0 plugins use the new Data Frame format to pass data from data sources to panels.
```
Comment (oddlittlebird) [inline docs/sources/developers/plugins/migration-guide.md:122]:
```suggestion
In <7.0, data source and panel plugins exchanged data using either time series or tables. In >=7.0 plugins use the new data frame format to pass data from data sources to panels.
```

File: docs/sources/developers/plugins/migration-guide.md [hunk 18]
```diff
@@ -0,0 +1,151 @@
...
+
+Grafana 7 is backward compatible with the old data format used in previous versions. Panels and data sources using the old format will still work with plugins using the new data frame format.
```
Comment (oddlittlebird) [inline docs/sources/developers/plugins/migration-guide.md:124]:
```suggestion
Grafana 7 is backward-compatible with the old data format used in previous versions. Panels and data sources using the old format still work with plugins using the new data frame format.
```

File: docs/sources/developers/plugins/migration-guide.md [hunk 19]
```diff
@@ -0,0 +1,151 @@
...
+      const timeSeries: TimeSeries = await doLegacyRequest(query);
+      return toDataFrame(timeSeries);
+    }
+  };
+}
+```
+
+For more information, refer to [Data frames](https://grafana.com/docs/grafana/latest/developers/plugins/data-frames/).
+
+## Troubleshooting
```
Comment (oddlittlebird) [inline docs/sources/developers/plugins/migration-guide.md:147]: Avoid stacked headings. Add at least a one-sentence intro to tell the user what to expect in this section.

Also, please make the heading more specific, like "Troubleshoot migration." I'm thinking of all the search results called "Troubleshooting" that will come up if we have multiple "Troubleshooting" headings in the docs.

File: docs/sources/developers/plugins/migration-guide.md [hunk 20]
```diff
@@ -0,0 +1,151 @@
...
+
+### Why doesn't my plugin show up after upgrading to Grafana 7?
```
Comment (oddlittlebird) [inline docs/sources/developers/plugins/migration-guide.md:149]: If possible, have more than one subheading per section.

File: docs/sources/menu.yaml [hunk 21]
```diff
@@ -481,6 +481,8 @@
           link: /developers/plugins/share-a-plugin/
         - name: Error handling
           link: /developers/plugins/error-handling/
+        - name: Migration guide
```
Comment (oddlittlebird) [inline docs/sources/menu.yaml:484]:
```suggestion
        - name: Plugin migration guide
```

**Review by oddlittlebird — CHANGES_REQUESTED** [review]
Great content!

**Review by oddlittlebird — APPROVED** [review]
LGTM. Great doc, Marcus!
